pipeline.cpp

```diff
@@ -82,6 +94,39 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
 void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
 {
     // Your code goes here
+    target[(int)triangle[0].y][(int)triangle[0].x] = attrs[0].color;
+    target[(int)triangle[1].y][(int)triangle[1].x] = attrs[1].color;
+    target[(int)triangle[2].y][(int)triangle[2].x] = attrs[2].color;
+
+
+    /* get the bounding box of the triangle */
+    int maxX = std::max(triangle[0].x, std::max(triangle[1].x, triangle[2].x));
+    int minX = std::min(triangle[0].x, std::min(triangle[1].x, triangle[2].x));
+    int maxY = std::max(triangle[0].y, std::max(triangle[1].y, triangle[2].y));
+    int minY = std::min(triangle[0].y, std::min(triangle[1].y, triangle[2].y));
+
+    /* spanning vectors of edge (v1,v2) and (v1,v3) */
+    Vertex vs1 = {triangle[1].x - triangle[0].x, triangle[1].y - triangle[0].y};
+    Vertex vs2 = {triangle[2].x - triangle[0].x, triangle[2].y - triangle[0].y};
+
+    for (int x = minX; x <= maxX; x++)
+    {
+        for (int y = minY; y <= maxY; y++)
+        {
+  
+            Vertex q = {x - triangle[0].x, y - triangle[0].y};
+
+            float s = (float)crossProduct(q.x,q.y, vs2.x,vs2.y) / crossProduct(vs1.x,vs1.y, vs2.x,vs2.y);
+            float t = (float)crossProduct(vs1.x,vs1.y, q.x, q.y) / crossProduct(vs1.x,vs1.y, vs2.x,vs2.y);
+
+            if ( (s >= 0) && (t >= 0) && (s + t <= 1))
+            { 
+                /* inside triangle */
+                Vertex q = {x,y};
+                DrawPrimitive(POINT,target,&q,attrs);
```

Comment:
Looks good, good use of the reading example and changing the methods and names to fit our project